### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by dfdx**
Main changes: 

1. Fitting process is split into 2 clear parts - gradient calculation and its application to parameters. By default, for this purpose functions `gradient_classic()` and `update_classic!()` are used. 
2. Classic updater is split into basic `update_weights!` and a number of functions modifying gradient, including those for weight decay (L1 and L2), sparsity, learning rate, etc. All such functions have the same signature, so one can easily combine them in an arbitrary way. 
3. Everything is configurable. Most functions take a "context" (`ctx`) - a dictionary used for holding configuration and buffers. We don't need to pass 3-5 options + buffers to each function anymore, now they all are combined into a context. 
4. In addition to configurable gradient and updater functions, user can now provide (via context) a scorer for measuring model "fitness" (by default, pseudo-likelihood is used) and a reporter to report progress during training (by default, `TextReporter` is used which simply prints current epoch, score and elapsed time). 
5. The package now supports both - `Float64` and `Float32`. Support for other (real?) data types may be provided given that BLAS functions are defined for them. 
6. API is mostly the same, little to no effort should be needed to start using the new version. 

Two things left outside this refactoring. One is EMF approximation by @sphinxteam which seems like a perfect option for gradient calculation. Corresponding paper has a little bit too much physics  for me, but if the team finds it possible to transform their work into this repository, it will be a great advantage for the whole Julia stats/deep learning community. 

Another thing that I unsuccessfully tried to borrow is visual monitor from the same fork. Unfortunately, in its current state it's closely bound to TAP free energy, which makes it quite hard to use as is. Yet, it's quite possible I will include something similar later. 

Comments and criticism are highly welcome. 

CC: @Rory-Finnegan @eric-tramel

### CODE REVIEW

File: src/conditional.jl [hunk 1]
```diff
@@ -1,59 +1,67 @@
...
-
-function Base.show{V,H}(io::IO, rbm::RBM{V,H})
-    n_vis = size(rbm.vbias, 1)
-    n_hid = size(rbm.hbias, 1)
-    print(io, "RBM{$V,$H}($n_vis, $n_hid)")
-end
-
-function split_vis(crbm::ConditionalRBM, vis::Mat{Float64})
+                        steps=5, sigma=0.01)
+    ConditionalRBM(Float64, Bernoulli, Bernoulli, n_vis, n_hid, steps;
```
Comment (Rory-Finnegan) [inline src/conditional.jl]: I think you want this to be `ConditionalRBM(Float64, V, H, n_vis, n_hid, ...`? If so I can change that when I add my generalize changes.

Comment (dfdx) [inline src/conditional.jl]: Correct, thanks!

File: src/rbm.jl [hunk 2]
```diff
@@ -1,118 +1,107 @@
...
     for j=1:size(means, 2), i=1:size(means, 1)
-        samples[i, j] = rand(Normal(means[i, j], sigma2))
+        samples[i, j] = T(rand(Normal(means[i, j], sigma2)))
     end
     return samples
 end
 
 
-function sample_hiddens{V,H}(rbm::AbstractRBM{V, H}, vis::Mat{Float64})
+function sample_hiddens{T,V,H}(rbm::AbstractRBM{T,V,H}, vis::Mat{T})
```
Comment (Rory-Finnegan) [inline src/rbm.jl]: Might want to accept a `vis` matrix of type`F<:AbstractFloat` and do a `convert(Array{T}, vis)` if `F` != `T`

Comment (dfdx) [inline src/rbm.jl]: If we agree on implicit conversion, I don't see a reason to limit `T` to be an instance of `AbstractFloat` - one may be interested in passing matrix of `Int`s or even `Boolean`, which is very natural for Bernoulli RBM.
